src/Config/Converters/EntitySourceConverterFactory.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Azure.DataApiBuilder.Config.ObjectModel;
+
+namespace Azure.DataApiBuilder.Config.Converters;
+
+internal class EntitySourceConverterFactory : JsonConverterFactory
+{
+    /// <inheritdoc/>
+    public override bool CanConvert(Type typeToConvert)
+    {
+        return typeToConvert.IsAssignableTo(typeof(EntitySource));
+    }
+
+    /// <inheritdoc/>
+    public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
+    {
+        return new EntitySourceConverter();
+    }
+
+    private class EntitySourceConverter : JsonConverter<EntitySource>
+    {
+        public override EntitySource? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType == JsonTokenType.String)
+            {
+                string? obj = reader.DeserializeString();
+                return new EntitySource(obj ?? "", EntitySourceType.Table, new(), Array.Empty<string>());
+            }
+
+            // Remove the converter so we don't recurse.
+            JsonSerializerOptions innerOptions = new(options);
+            innerOptions.Converters.Remove(innerOptions.Converters.First(c => c is EntitySourceConverterFactory));
+
+            EntitySource? source = JsonSerializer.Deserialize<EntitySource>(ref reader, innerOptions);
+
+            if (source?.Parameters is not null)
+            {
+                // If we get parameters back the value field will be JsonElement, since that's what STJ uses for the `object` type.
+                // But we want to convert that to a CLR type so we can use it in our code and avoid having to do our own type checking
+                // and casting elsewhere.
+                return source with { Parameters = source.Parameters.ToDictionary(p => p.Key, p => GetClrValue((JsonElement)p.Value)) };
+            }
+
+            return source;
+        }
+
+        private static object GetClrValue(JsonElement element)
+        {
+            return element.ValueKind switch
+            {
+                JsonValueKind.String => element.GetString() ?? "",
+                JsonValueKind.Number => element.GetInt32(),
```

Comment:
Given that it's coming from JSON the value is really an IEEE-754 double precision floating point number (see https://en.wikipedia.org/wiki/JSON#Data_types).

Given the usage here we should be reasonably confident on the type being a valid `int`, but if we're not we should expand the test coverage to represent real world values we're seeing so we can handle it accordingly.

Reply:
These parameters are default values of stored procedure parameters which could be decimal or int - unless we know what the underlying parameter type this is - we cannot cast it one way or the other. So, looks like the value in the config is always cast to a string before converting to a GraphQLType at time of schema creation. 

See here:
https://github.com/Azure/data-api-builder/blob/0b3e3abf6085902fac03a4d8dee6d17e9d760533/src/Service.GraphQLBuilder/GraphQLStoredProcedureBuilder.cs#LL48C21-L48C21

We do have tests that assume these parameters are string values but these are unit tests that dont read from the config:
https://github.com/Azure/data-api-builder/blob/0b3e3abf6085902fac03a4d8dee6d17e9d760533/src/Service.Tests/GraphQLBuilder/Sql/StoredProcedureBuilderTests.cs#L84

Reply:
Ah, I see how the test is bypassing the serializer and that's how I missed them.

I'll have to think of how we can ensure the test coverage is there.

Reply:
Test coverage would be a good follow up PR to address the pre-existing gap.

To address Ani's initial comment, Aaron added `GetNumberValue()` which "_attempts to get the correct numeric value from the JsonElement, If all possible numeric values are exhausted, the raw text is returned._" and attempts to parse Int32, Decimal, Double, Int64, UInt32/64, Single, -> or raw text (string).

This can be resolved